src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs

```diff
@@ -0,0 +1,399 @@
+    ///
+    /// For use when initializing another [`TimestampOracle`] implementation
+    /// from another oracle's state.
+    ///
+    /// NOTE: This method can have linearizability violations, but it's hard to
+    /// get around those. They will only occur during bootstrap, and only
+    /// if/when we migrate between different oracle implementations. Once we
+    /// migrated from the catalog-backed oracle to the postgres/crdb-backed
+    /// oracle we can remove this code and it's callsite.
+    pub(crate) async fn get_all_timelines(
+        config: PostgresTimestampOracleConfig,
+    ) -> Result<Vec<(String, Timestamp)>, anyhow::Error> {
+        let postgres_client = PostgresClient::open(config.into())?;
+
+        let client = postgres_client.get_connection().await?;
+
+        let q = r#"
+            SELECT EXISTS (SELECT * FROM information_schema.tables where table_name = 'timestamp_oracle');
+        "#;
+        let statement = client.prepare(q).await?;
+        let exists_row = client.query_one(&statement, &[]).await?;
+        let exists: bool = exists_row.try_get("exists").expect("missing exists column");
+        if !exists {
+            return Ok(Vec::new());
+        }
+
+        let q = r#"
+            SELECT timeline, GREATEST(read_ts, write_ts) as ts FROM timestamp_oracle;
+        "#;
+        let statement = client.prepare(q).await?;
+        let rows = client
+            .query(&statement, &[])
+            .await
+            .expect("todo: add internal retries");
+
+        let result = rows
+            .into_iter()
+            .map(|row| {
+                let timeline: String = row.try_get("timeline").expect("missing timeline column");
+                let ts: i64 = row.try_get("ts").expect("missing ts column");
+                let ts = Timestamp::try_from(ts).expect("we don't use negative timestamps");
+
+                (timeline, ts)
+            })
+            .collect::<Vec<_>>();
+
+        Ok(result)
+    }
+
+    #[tracing::instrument(name = "oracle::write_ts", level = "debug", skip_all)]
+    async fn fallible_write_ts(&self) -> Result<WriteTimestamp<Timestamp>, anyhow::Error> {
+        let proposed_next_ts = self.next.now();
+        let proposed_next_ts =
+            i64::try_from(proposed_next_ts).expect("timestamp does not fit into i64");
+
+        let q = r#"
+            UPDATE timestamp_oracle SET write_ts = GREATEST(write_ts+1, $2)
+                WHERE timeline = $1
+            RETURNING write_ts;
+        "#;
+        let client = self.get_connection().await?;
+        let statement = client.prepare_cached(q).await?;
+        let result = client
+            .query_one(&statement, &[&self.timeline, &proposed_next_ts])
+            .await?;
+
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "write_ts");
```

Comment:
Maybe something more descriptive like "returning write_ts", here and elsewhere. I'd also include any arguments to the query in the log message

Reply:
will do :ok_hand: